### PR TITLE
Handle the case of multiple IP addresses for publicDNSIP

### LINKED ISSUE #2180: OCP 4.12: LB service on AWS returns multiple IPs and cause to DP failure
**What happened**:

Deployed Submariner using LB service on OCP-4.12.0-ec.5 (CNI=OVNKubernetes) on AWS

IPSec tunnels are deleted and recreated periodically  , check [1]

**How to reproduce it (as minimally and precisely as possible)**:
Deploy Submariner with LB on OCP-4.12.0-ec.5 (CNI=OVNKubernetes), cloud=AWS
**Anything else we need to know?**:
Seems that LB external IP [2]  resolved to multiple IP addresses,
and when the LB service is associated with multiple IPs, every time we make a query the order in which the IPs are returned changes [3], so public-ip watcher thinks that the public-ip is modified.

[1] 
[90m2022-12-07T09:09:43.439Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:10:23.487Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:12:03.597Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:13:03.663Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:13:43.711Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:14:03.737Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:15:43.834Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:16:23.882Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:16:43.909Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:17:03.942Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:17:23.972Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:17:44.003Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:18:04.027Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:18:44.077Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:19:44.142Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:20:04.170Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:20:24.200Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:21:04.246Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:21:24.289Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:21:44.325Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:22:04.351Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:22:44.399Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:23:04.426Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:23:44.476Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:25:24.586Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:25:44.617Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:26:04.645Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:26:24.676Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:26:44.706Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:27:24.757Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:28:24.817Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:28:44.844Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:30:04.955Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:30:44.997Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:31:05.029Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:31:25.060Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:31:45.090Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:32:05.118Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:32:25.148Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:32:45.178Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:33:05.211Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "13.58.40.253"
[90m2022-12-07T09:33:25.244Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:33:45.273Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:34:05.302Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"
[90m2022-12-07T09:34:25.327Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.191.84.174"
[90m2022-12-07T09:35:25.392Z[0m [32mINF[0m ..public_ip_watcher.go:77 Endpoint             Public IP changed for the Gateway, updating the local endpoint with publicIP "18.117.41.113"

[2]
$ oc --kubeconfig ~/Downloads/kubeconfig-cluster2  describe  svc -n submariner-operator  submariner-gateway 

Name:                     submariner-gateway
Namespace:                submariner-operator
Labels:                   <none>
Annotations:              service.beta.kubernetes.io/aws-load-balancer-type: nlb
Selector:                 app=submariner-gateway,gateway.submariner.io/status=active
Type:                     LoadBalancer
IP Family Policy:         SingleStack
IP Families:              IPv4
IP:                       172.30.243.103
IPs:                      172.30.243.103
LoadBalancer Ingress:     a2bce4344a2a14fef97974e070f7eb72-1512d4312814df40.elb.us-east-2.amazonaws.com
Port:                     cable-encaps  4500/UDP
TargetPort:               4500/UDP
NodePort:                 cable-encaps  32766/UDP
Endpoints:                10.0.146.210:4500
Port:                     natt-discovery  4490/UDP
TargetPort:               4490/UDP
NodePort:                 natt-discovery  31784/UDP
Endpoints:                10.0.146.210:4490
Session Affinity:         None
External Traffic Policy:  Local
HealthCheck NodePort:     31681
Events:
  Type    Reason                Age   From                Message
  ----    ------                ----  ----                -------
  Normal  EnsuringLoadBalancer  132m  service-controller  Ensuring load balancer
  Normal  EnsuredLoadBalancer   132m  service-controller  Ensured load balancer

[3]

$ dig a2bce4344a2a14fef97974e070f7eb72-1512d4312814df40.elb.us-east-2.amazonaws.com +short
18.117.41.113
18.191.84.174
13.58.40.253
$ dig a2bce4344a2a14fef97974e070f7eb72-1512d4312814df40.elb.us-east-2.amazonaws.com +short
13.58.40.253
18.117.41.113
18.191.84.174

### PR DESCRIPTION
**Pull request by yboaron**
We have noticed that sometimes on AWS platform the ingress.hostname of LB service is resolved to multiple IP addresses, and each time their order may change.

This PR updates PublicDNSIP to support multiple IP addresses.

Fixes: https://github.com/submariner-io/submariner/issues/2180

Signed-off-by: Yossi Boaron <yboaron@redhat.com>

### CODE REVIEW

File: pkg/endpoint/public_ip.go [hunk 1]
```diff
@@ -212,6 +214,12 @@ func publicDNSIP(clientset kubernetes.Interface, namespace, fqdn string) (string
 		return "", errors.Wrapf(err, "error resolving DNS hostname %q for public IP", fqdn)
 	}
 
+	if len(ips) > 1 {
+		sort.Slice(ips, func(i, j int) bool {
```
Comment (sridhargaddam) [inline pkg/endpoint/public_ip.go:219]: I think we have to understand the reason for multiple IPs on AWS (in OCP4.12 and not in OCP4.11) and if we can successfully connect with ALL the IPs returned by DNS query. If so, this approach should work.

**Comment by sridhargaddam** [comment]
Looks good @yboaron. Do you want to add a unit-test case for this use-case?

**Comment by yboaron** [comment]
> Looks good @yboaron. Do you want to add a unit-test case for this use-case?

@sridhargaddam  Any idea how to simulate multi IP addresses DNS Resolution ?  
IIUC, [multi IP address is not supported in nip.io ](https://github.com/exentriquesolutions/nip.io/issues/23)

**Comment by nyechiel** [comment]
@yboaron how critical this fix is? should we backport it?

**Comment by yboaron** [comment]
> @yboaron how critical this fix is? should we backport it?

@nyechiel well, LB service is relevant only for ROSA and ARO, I couldn't even reproduce the issue on AWS, in my tests LB DNS hostname resolved to more than one IP only for a short time about 30-60 seconds,

I don't think we should backport this PR

**Comment by sridhargaddam** [comment]
Personally I have noticed this behaviour only recently and I'm not sure if AWS is experimenting this feature.
Either way, as long as the public IPs associated with ELB falls back to a single IP (hopefully in a minute or two as Yossi observed) which points to the zone in which submariner-gateway pod is scheduled, we should be good.

**Comment by sridhargaddam** [comment]
> > Looks good @yboaron. Do you want to add a unit-test case for this use-case?
> 
> @sridhargaddam Any idea how to simulate multi IP addresses DNS Resolution ? IIUC, [multi IP address is not supported in nip.io ](https://github.com/exentriquesolutions/nip.io/issues/23)

Correct Yossi, that's what I'm seeing too. Thanks for checking.

**Review by sridhargaddam — APPROVED** [review]

**Review by skitt — APPROVED** [review]

**Pull request merged**